Framework: Implemented data-layer freshness verifier (take 2)

jorgefilipecosta opened the pull request:
The PR is the take 2 of https://github.com/Automattic/wp-calypso/pull/16728, it creates a readme file that documents a possible fresh verification approach based on the original idea and the feedback received.

dmsnell commented:
I'm somewhat wondering how much we should put into polling in places like this vs getting our push-connections established more quickly. For example, would our efforts be better spend not on adding auto-polling into Calypso but instead on supporting server-side push for the Reader and for shopping carts etc… I think the mechanisms mainly exist already but we need to establish a pattern, some documentation, a guide, and some norms…

jorgefilipecosta commented:
Hi @dmsnell 

> what identifies a resource and distinguishes one request from another?
> can we opt out? should we?
> is there a sensible default across all objects? is there one that I'm unaware of?
Very important questions I will try to add answers in the readme.

> what happens when I have two components, one of which requires a freshness of five minutes and another of ten seconds, and the ten-second component goes away? when will the data continue to update?
> how do we indicate that we should continue updating vs. simply needing fresh data?

These questions are related with polling, in my vision data-layer does not offer polling, data-layer just offers the mechanism to avoid requests if they are fresh, but never re-requests them if no new request is made.

If polling is really needed, we can add Interval component on the query component, that periodically issues requests. The freshness mechanism of data-layer avoids the number of requests being multiplied by the number of-query components on the page because duplicate requests are removed.
So answering based on this idea:

> what happens when I have two components, one of which requires a freshness of five minutes and another of ten seconds, and the ten-second component goes away? when will the data continue to update?
The components that needs freshness of five seconds, each five seconds dispatches a request for data and when data is not five seconds fresh the request is made. The one of ten seconds freshness each 10 seconds dispatches a request for data and if data is more than seconds old it requests it (never happens after maybe the first time because we have 5 seconds requester). If the 10 seconds component goes away the 5 seconds one continues to request and get is freshness as before nothing changes.


In my option freshness is not a mechanism for polling and does not promote polling it just makes it easy to do if really needed by combining freshness with the interval component. I think the best way to add this answers is to write a short mini polling approach description but saying that polling should not be used unless in the presence of a very specific case where we need fresh data and our push mechanism can not handle (e.g specific external service). It may also make sense to add that part to the interval component.

> I'm somewhat wondering how much we should put into polling in places like this vs getting our push-connections established more quickly.
I'm in favor of push-connections and not spending resources on polling. This freshness mechanism is used to avoid duplicate requests like in rendering/unrendering/rendering and similar situations and I think it has value even if not used in polling. It makes polling easy if used together with the interval component, but we should not be adding this to query components right now unless a very specific situation arises. The efforts in data continuously updated scenarios should go for a the push-solution. If in the future the push solution does not handle a specific case we can use the mix  (interval + freshness).

dmsnell commented:
All good stuff @jorgefilipecosta! I was prodding with those questions mainly as a way to suggest that people might be asking them and the README is the right place to answer them in anticipation :wink:

On scoping: Yeah I think leaving out polling from the data layer right now is a wonderful decision. 

On blocking old requests: an interesting take on it upside-down from how I've seen it. I've seen it was a mechanism whereby we can allow the framework to make a contract with components that if they need data _at least as recently-updated as requested_ then it will happen. I guess it's also valid to frame it as _unless the data is possibly staler than you need then I won't let you refresh it_.

jorgefilipecosta commented:
Hi @dmsnell, I tried to add answers to the question you raised and I removed implementation details section. To answer questions related with polling I added a question "A possible polling implementation", but I clearly stated polling is not recommended, I'm not certain if this section should exist but I was not finding other way to answer the questions you raised (related with polling).

> I've seen it was a mechanism whereby we can allow the framework to make a contract with components that if they need data at least as recently-updated as requested then it will happen.

You mean allow components to pass a timestamp to the data-layer and if the last request was after that time it does not happen if it was before (or never executed) it happens?

dmsnell commented:
> You mean allow components to pass a timestamp to the data-layer and if the last request was after that time it does not happen if it was before (or never executed) it happens?

not a timestamp, but a length of time which means, "if the latest data is older than (now - length of time) then make sure you try and refetch it"

in effect there's nothing different than how you built it I think; just a positive vs. a negative description of the process

dmsnell commented on client/state/data-layer/freshness-verifier/README.md:
```diff
@@ -0,0 +1,45 @@
+ siteId: 12886750,
+ postId: 97,
+ ...
+ freshness: 2000
```
I think this should definitely live in action meta data under `meta.dataLayer.fresherThan` or `meta.dataLayer.freshness` so that we don't conflate it with the properties of the action itself

jorgefilipecosta replied:
Corrected 👍

dmsnell commented on client/state/data-layer/freshness-verifier/README.md:
````diff
@@ -0,0 +1,45 @@
+```
+When a request is still fresh it does not trigger a network request and is ignored.
+
+The freshness mechanism is optional and if no freshness is set in the handler or the action the freshness mechanism is not executed. If a freshness value is set on the handler but in a given component freshness should not be used it's possible to opt-out its usage by setting `freshness: 0` in the action being dispatched. Opt-out of the mechanism is not recommended unless in the presence of a spec ...
````
We might want to explain that this isn't exactly _opt-out_ (or at least, I don't think it is) but rather the logical way to demand that data be as new as possible. That is, if it's older than zero milliseconds refresh it!

jorgefilipecosta replied:
Hi @dmsnell, I think there are 2 different cases, one where we set freshness in dispatchRequest but on the action, we set freshness = 0 and in fact, that's not opt-out and the text was corrected. But the other case is when we don't set freshness in the action and in the dispatchRequest call. In that case, in my option, we have opt-out because the freshness mechanism does nothing, it just calls the normal functionality as if freshness did not exist e.g. if( ! freshness ) { return initiator(..)}. The text was corrected to clearly present this assumption. Assuming 0 freshness when freshness is not defined is also valid the only disadvantage is that we would be saving the request last execution when it is not needed. Let me know if you prefer some change.

dmsnell commented on client/state/data-layer/freshness-verifier/README.md:
```diff
@@ -0,0 +1,45 @@
+
+The freshness mechanism is optional and if no freshness is set in the handler or the action the freshness mechanism is not executed. If a freshness value is set on the handler but in a given component freshness should not be used it's possible to opt-out its usage by setting `freshness: 0` in the action being dispatched. Opt-out of the mechanism is not recommended unless in the presence of a spec ...
+
+The default value for the freshness of the actions of a given type is set in the handler so if the action does not set its own freshness value the one set in the handler of the request is used. There is no global default value if the both handler and action don't set a freshness value, freshness is not used.
```
This paragraph might make more sense in an ordered list of where we get the values…

```md
The freshness for a specific request action is determined in the following priority:
  1. from action metadata at `meta.dataLayer.freshness`
  2. from the optional `freshness` parameter in the call to `dispatchRequest()` in the data layer handler for the action type
  3. a globally-assumed default of `0`, implying that requests will be instantly fetched as often as they dispatch
```

dmsnell commented on client/state/data-layer/freshness-verifier/README.md:
```diff
@@ -0,0 +1,45 @@
+
+The default value for the freshness of the actions of a given type is set in the handler so if the action does not set its own freshness value the one set in the handler of the request is used. There is no global default value if the both handler and action don't set a freshness value, freshness is not used.
+
+In order to distinguish one request from another, the data-layer uses information of all the properties in the request action excluding meta and freshness properties. So if two request actions have all properties equal they are assumed to be the same request. If there is a difference in the properties, they are assumed to be different requests. Internally the data-layer converts the action to a k ...
```
> excluding meta _and_ freshness properties

see note about about putting freshness information in metadata

dmsnell replied:
> Internally the data-layer converts the action to a key (string) and then uses that keys to compare requests. 

probably not important here :smile:

dmsnell commented on client/state/data-layer/freshness-verifier/README.md:
```diff
@@ -0,0 +1,45 @@
+
+The polling solution consists in adding the interval component to the render function of the query component where we need periodic requests. The interval component is used to periodically dispatch an action request with the freshness needed. The freshness mechanism avoids the number of requests being multiplied by the number of query components on the page because duplicate requests are removed.
+If there exists more than one polling component and one component is unmount the polling continues because the other component continues to issue requests.
+It is possible to safely add a "polling" query component to the places that need that fresh data. When the data is not needed by any component the requests will stop being executed.
```
How about an example of a  basic query component implementation with polling?

I don't think we need to warn so strongly against polling here since we don't have great alternatives at the moment.

jorgefilipecosta commented:
Hi @dmsnell, thank you for your efforts in reviewing this. I tried to address your feedback.

One remaining discussion point is what happens to the request actions that are discarded.
Two options exist:
- The data-layer middleware discards it and avoids that it arrives in reducers.
This one makes the data-layer more complex but has the advantage of being compatible some existing request actions that when they arrive in the reduces the reducer set a isRequesting state, and that state is used in components e.g. to not show something to show placeholders etc..

- The data-layer allows the action to arrive in the reducer.
In this situation, I think in the dispatch action call besides passing freshness we should allow handlers to pass an onFreshRequest function that can be used to dispatch actions that clean the isRequesting state.

jorgefilipecosta commented:
This PR is On Hold untill https://github.com/Automattic/wp-calypso/pull/17263 is merged.

coderkevin commented:
Just got back from AFK, will look at this tomorrow. Thanks for the continued work, @jorgefilipecosta

jorgefilipecosta commented:
Hi,
Given the work being done in PR https://github.com/Automattic/wp-calypso/pull/17263 this PR is changed to On Hold. After the work on the related PR is concluded we get a better idea of how our freshness functionality will be and how it will interact with the data-layer. If needed the read me will be updated in accordance with the new meta data functionality tracker offered.

dmsnell commented:
Note that with the merge of #17263 we have some new foundations laid that might make this task easier to achieve now.

jorgefilipecosta commented:
Hi @dmsnell, thank you a lot for the effort in landing https://github.com/Automattic/wp-calypso/pull/17263. It will for sure make freshness easier to implement. As soon as I can, I will try to update this PR to make use of the new mechanism.

apeatling commented:
@jorgefilipecosta It's been a long time with these open, so what I'm going to do is close all of your older PRs, and we'll (calypso team) take a look at them and either re-open or reuse some of the work as part of our upcoming projects.
